jpa/eclipselink.jpa.test/src/org/eclipse/persistence/testing/tests/jpa/advanced/PersistenceUnitProcessorTest.java

```diff
@@ -72,6 +72,34 @@ protected URLConnection openConnection(URL u) throws IOException {
                         "META-INF/persistence.xml"
                 ).toString()
         );
+
+        Assert.assertEquals(
+            // The protocol changes - we view ZIPs as JARs.
+            "jar:file:/foo/bar.war!/WEB-INF/classes",
+            PersistenceUnitProcessor.computePURootURL(
+                new URL("zip", "", -1, "/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml", dummyZipHandler), 
+                "META-INF/persistence.xml"
+            ).toString()
+        );
+    }
+
+    public void testComputePURootURLForJarFile() throws Exception {
+
+        Assert.assertEquals(
+            new URL("file:/foo/bar.jar"),
+            PersistenceUnitProcessor.computePURootURL(
+                new URL("jar:file:/foo/bar.jar!/META-INF/persistence.xml"),
+                "META-INF/persistence.xml"
+            )
+        );
+
+        Assert.assertEquals(
+            new URL("jar:file:/foo/bar.war!/WEB-INF/classes"),
+            PersistenceUnitProcessor.computePURootURL(
+                new URL("jar:file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml"),
+                "META-INF/persistence.xml"
+            )
+        );
```

Comment:
Perhaps you should take a look at the implementation for ```wsjar```:
https://github.com/eclipse-ee4j/eclipselink/blob/ea5c77487819d44b821bee545651d85a8cd93535/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java#L199-L219

For WebSphere, we rewrap our ```jar``` protocol with wsjar. This is why we added a special case here for just this protocol, but it seems the implementation tried to correctly take into account this usecase:

``` java
                // If this doesn't reference a war file with a properly located persistence.xml,
                // then chop off everything after the "!/" marker and assume it is a normal jar.
                // Else, if the wsjar URL references a file with a ".war" extension, and its entry
                // starts with WEB-INF/classes/, then the calculated persistence unit root should
                // be wsjar:path/to/a.war!/WEB-INF/classes/ as per JPA 2.1 Spec section 8.2 "Persistence Unit Packaging".
                // Filter out invalid scenarios such as wsjar:file:/a/path/to/my.war!/foo/WEB-INF/classes/META-INF/persistence.xml
```
This may explain why I have not seen this as a problem :/

Reply:
> I'm surprised we do not correctly calculate the root for this.

> This may explain why I have not seen this as a problem :/

We've been dealing with Tomcat the whole time, which explodes jars by default - I'm guessing other web containers are doing the same thing?.. I've only stumbled upon this issue once I attempted to use embedded Jetty for this (and package everything into a jar - running from a directory is OK).

Reply:
> Should I just reuse this (restrictive) logic? The way I see it, zip, jar and wsjar should all work the same.
Not sure what you mean by "restrictive"... are you referring to the fact that we only want to support specification supported URLs?

Reply:
> Not sure what you mean by "restrictive"... are you referring to the fact that we only want to support specification supported URLs or the fact that we have each protocol with separate logic?

By "restrictive" I meant the fact that we go out of our way to make sure that any cases not conforming to the specification (like `jar:file:foo.jar!/bar/META-INF/persistence.xml`) resolve to the top of the archive. The thing that irks me most about this approach is that resolution works differently depending on the packaging:
```
jar:file:/foo.jar!/bar/META-INF/persistence.xml -> file:/foo.jar
file:/foo/bar/META-INF/persistence.xml          -> file:/foo/bar
```
This can lead to all sorts of subtle and confusing bugs, and this is what bit me in the first place. Finding some configuration that maps the same way in both cases works (somewhat) in my case, but I still find this behavior surprising. Look at it this way: since these URLs don't conform to the spec, we have a choice to either raise an error, or handle them in the most consistent and predictable way possible. As I see it, the existing approach of using the archive root fails to do this.

However since we're keeping this logic for `wsjar` anyway, I guess it does make a modicum of sense to do the same thing for other archives as well. At least for my case fixing `WEB-INF/classes` should be good enough.

Reply:
>By "restrictive" I meant the fact that we go out of our way to make sure that any cases not conforming to the specification (like jar:file:foo.jar!/bar/META-INF/persistence.xml) resolve to the top of the archive.

Maybe we should throw some kind of warning if the user has their persistence.xml at a bad location. We can then proceed forward, but the warning will have let the user know we are operating off a bad location.

> we have a choice to either raise an error, or handle them in the most consistent and predictable way possible. As I see it, the existing approach of using the archive root fails to do this.

I think I'd rather raise an error. So, if I understand your situation correctly, the fact that the archive root is being returned is causing the issue later in EclipseLink? We set the URL to point at the archive root, like ```path/to/foo.jar```, and this can't be handled correctly somewhere in EclipseLink?

Reply:
> I think I'd rather raise an error.

Personally, I'd rather have consistent behavior in this cases, but for a reference implementation an error is also OK. I'll figure something out on my end.

> So, if I understand your situation correctly, the fact that the archive root is being returned is causing the issue later in EclipseLink? We set the URL to point at the archive root, like path/to/foo.jar, and this can't be handled correctly somewhere in EclipseLink?

As I recall, the error itself was pretty simple - the "missing" `persistence.xml` resulted in a `@NamedQuery` lookup issue. Sure, there was a warning somewhere (I think), but spotting it in a log of a large-ish application is not easy, especially if you don't know what to look out for. The confusing part was that everything worked OK in the development environment (when running from an exploded directory), but failed for the JAR.